### ISSUE TITLE
Dosbox mapper blue not white

### ISSUE THREAD

**Issue by hex007** [RetroPie/RetroPie-Setup#1952]
After a clean install of Retropie 4.2 on Raspberry pi zero, I installed DOSBOX from source. The Mapper `CTRL+F1` looks like this. Connected to HDMI to take picture but I will be using it on 3.5 inch display. The blue text is unreadable. Mouse is above `ESC` button

![img_20170415_000904](https://cloud.githubusercontent.com/assets/1689246/25068327/be6cbaaa-2214-11e7-97b7-6546de996b8f.jpg)

**Comment by hex007** [comment]
I can verify that this also affects Raspberry Pi 3b

**Comment by joolswills** [comment]
You already have a topic open on the forum -I prefer not to have issues opened in multiple places (info when opening a ticket asks to use the forum first - it doesn't mean open a topic, then also post here). This is likely a DOSBOX or SDL issue rather than being RetroPie specific.

Is this your ticket also ? https://sourceforge.net/p/dosbox/bugs/452/

**Comment by hex007** [comment]
Yes. Since i wasnt sure where the problem was from, i opened this issue before diving into the code.

**Comment by hex007** [comment]
@joolswills Can you try running this program. [https://pastebin.com/7SQZiPcH](url)

I was able to isolate the problem quite a bit. This program creates a grey scale gradient on my laptop running Fedora 25 with SDL from libsdl website. It creates a series of colors on Raspberry pi 3 running Retropie 4.2. Screenshots below. 

Surprisingly the code runs properly on both Fedora and Retropie when LEVELS = 16

`SDL_SetPalette(screen, SDL_LOGPAL|SDL_PHYSPAL,  colors, 0, LEVELS);` 

If number of colors is not 16 the colors are not the same. This suggest SDL error on Retropie.

![screenshot from 2017-04-23 16-28-54](https://cloud.githubusercontent.com/assets/1689246/25318388/33deb5bc-2842-11e7-8147-43e2a0bd95db.png)

![img_20170423_162801](https://cloud.githubusercontent.com/assets/1689246/25318392/3a4f50fa-2842-11e7-98f9-41fa081f9dc3.jpg)

**Comment by joolswills** [comment]
Did you try switching dispmanx sdl2 backend on/off for dosbox to see if that helps?

I doubt any more work will be done on sdl1.x - it's more likely we will move to sdl2 for dosbox anyway.

I don't have any free time to look into this currently.

**Comment by hex007** [comment]
I have a solution for this. Should I submit a patch?

**Comment by joolswills** [comment]
what solution ? against sdl or dosbox ?

**Comment by hex007** [comment]
patch for the dosbox source. all you have to do is fill map_pal with blank colors so size goes from 6 to 16 and change corresponding SetPalette ncolors to 16. Diff below

`--- sdl_mapper.cpp	2017-04-24 15:34:46.721040544 -0700
+++ sdl_mapper.cpp	2017-04-24 15:40:45.407736264 -0700
@@ -2003,13 +2003,24 @@
 	bind_but.bind_title->Change("Bind Title");
 }

-static SDL_Color map_pal[6]={
+static SDL_Color map_pal[16]={
 	{0x00,0x00,0x00,0x00},			//0=black
 	{0x7f,0x7f,0x7f,0x00},			//1=grey
 	{0xff,0xff,0xff,0x00},			//2=white
 	{0xff,0x00,0x00,0x00},			//3=red
 	{0x10,0x30,0xff,0x00},			//4=blue
-	{0x00,0xff,0x20,0x00}			//5=green
+	{0x00,0xff,0x20,0x00},			//5=green
+	{0x00,0x00,0x00,0x00},			//blank
+	{0x00,0x00,0x00,0x00},			//blank
+	{0x00,0x00,0x00,0x00},			//blank
+	{0x00,0x00,0x00,0x00},			//blank
+	{0x00,0x00,0x00,0x00},			//blank
+	{0x00,0x00,0x00,0x00},			//blank
+	{0x00,0x00,0x00,0x00},			//blank
+	{0x00,0x00,0x00,0x00},			//blank
+	{0x00,0x00,0x00,0x00},			//blank
+	{0x00,0x00,0x00,0x00}			//blank
 };

 static void CreateStringBind(char * line) {
@@ -2362,7 +2373,7 @@
 	if (mapper.surface == NULL) E_Exit("Could not initialize video mode for mapper: %s",SDL_GetError());

 	/* Set some palette entries */
-	SDL_SetPalette(mapper.surface, SDL_LOGPAL|SDL_PHYSPAL, map_pal, 0, 6);
+	SDL_SetPalette(mapper.surface, SDL_LOGPAL|SDL_PHYSPAL, map_pal, 0, 16);
 	if (last_clicked) {
 		last_clicked->BindColor();
 		last_clicked=NULL;
`

**Comment by joolswills** [comment]
Please can you wrap it in a code block

**Comment by hex007** [comment]
Attached file instead of code block. The code block was getting broken due to text in diff.

**Comment by joolswills** [comment]
code block should work fine - I put diffs on github all the time. just use three backticks above and three below (can use ```diff for diff colorisation

**Comment by joolswills** [comment]
you used a single backtick which is why it didn't work btw.

**Comment by joolswills** [comment]
Thanks for the patch. Will look into it when I have time - did you try btw with dispmanx backend on/off without the patch ?

**Comment by hex007** [comment]
I dont know how to. Can you elaborate more.

**Comment by joolswills** [comment]
RetroPie-Setup -> Configuration  / Tools -> Configure emulators to use dispmanx SDL

Change the dosbox setting - restart dosbox from ES (not manually as the dispmanx setting is handled via runcommand).

**Comment by hex007** [comment]
Enabling dispmanx SDL works too. It is not enabled by default. 

Although there is no harm in adding 16 colours to palette instead of 6.

**Comment by joolswills** [comment]
Just to note it would probably be worth opening an upstream bug agaist sdl for this - I will look into including the code workaround, but if it's a sdl issue it should be reported upstream. https://bugzilla.libsdl.org/

**Comment by joolswills** [comment]
Sorry. I had a brain glitch - this is sdl1 still, so no point reporting upstream. I will look into the sdl1 code and see if I spot anything obvious. Could be a RPI framebuffer issue

**Comment by hex007** [comment]
@joolswills which patches are used to patch sdl1 on RPI
